### PR TITLE
[Merged by Bors] - Remove the `SystemParamState` trait and remove types like `ResState`

### PR DESCRIPTION
**Pull request by JoJoJet**
Spiritual successor to #5205.
Actual successor to #6865.

# Objective

Currently, system params are defined using three traits: `SystemParam`, `ReadOnlySystemParam`, `SystemParamState`. The behavior for each param is specified by the `SystemParamState` trait, while `SystemParam` simply defers to the state.

Splitting the traits in this way makes it easier to implement within macros, but it increases the cognitive load. Worst of all, this approach requires each `MySystemParam` to have a public `MySystemParamState` type associated with it.

## Solution

* Merge the trait `SystemParamState` into `SystemParam`.
* Remove all trivial `SystemParam` state types. 
  * `OptionNonSendMutState<T>`: you will not be missed.

---

- [x] Fix/resolve the remaining test failure.

## Changelog

* Removed the trait `SystemParamState`, merging its functionality into `SystemParam`.

## Migration Guide

**Note**: this replaces the migration guide for #6865.
This is relative to Bevy 0.9, not main.

The traits `SystemParamState` and `SystemParamFetch` have been removed, and their functionality has been transferred to `SystemParam`.

```rust
// Before (0.9)
impl SystemParam for MyParam<'_, '_> {
    type State = MyParamState;
}
unsafe impl SystemParamState for MyParamState {
    fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self { ... }
}
unsafe impl<'w, 's> SystemParamFetch<'w, 's> for MyParamState {
    type Item = MyParam<'w, 's>;
    fn get_param(&mut self, ...) -> Self::Item;
}
unsafe impl ReadOnlySystemParamFetch for MyParamState { }

// After (0.10)
unsafe impl SystemParam for MyParam<'_, '_> {
    type State = MyParamState;
    type Item<'w, 's> = MyParam<'w, 's>;
    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State { ... }
    fn get_param<'w, 's>(state: &mut Self::State, ...) -> Self::Item<'w, 's>;
}
unsafe impl ReadOnlySystemParam for MyParam<'_, '_> { }
```

The trait `ReadOnlySystemParamFetch` has been replaced with `ReadOnlySystemParam`.

```rust
// Before
unsafe impl ReadOnlySystemParamFetch for MyParamState {}

// After
unsafe impl ReadOnlySystemParam for MyParam<'_, '_> {}
```

### CODE REVIEW

**Comment by JoJoJet** [comment]
Before this can exit draft status, I'll need to fix up the remaining test failure. It is the doctest for the `SystemParamFunction` trait, defined here: https://github.com/bevyengine/bevy/blob/4820917af6be06866f10d4bd17191802d6640efb/crates/bevy_ecs/src/system/function_system.rs#L498

The reason that this test now fails is because since the `SystemParamState` trait was removed, we have lost the property that `SystemParamItem<SystemParamItem<T>>` == `SystemParamItem<T>`. Here's a few options for how to deal with this.

1. Find some way of restoring that property (no luck so far).
2. Accept that we lose that property, and delete/change that test accordingly.
3. Cancel this PR.
4. Something else I haven't thought of.

**Comment by JoJoJet** [comment]
### 1. Restore the property

So, we would like to restore the property that `SystemParamItem<SystemParamItem<T>>` == `SystemParamItem<T>`. The most obvious way of doing this is to add a bound to the `SystemParam::Item` associated type.

```rust
type Item<'w, 's>
    where for<'w, 's, 'a, 'b> Self::Item<'a, 'b>: SystemParam<Item<'w, 's> = Self::Item<'w, 's>>;
```

This (and similar variations) break the type recursion limit, and sometimes outright crash rustc.

**Comment by JoJoJet** [comment]
One workaround suggested by @BoxyUwU was to avoid generic associated types by defining the item on a new trait.

```rust
trait SystemParam: for<'w, 's> SystemParamGats<'w, 's> {}
trait SystemParamGats<'w, 's> {
    type Item;
}
```

So, instead of doing `T::Item<'w, 's>`, you would do `<T as SystemParamGats<'w, 's>>::Item`. Theoretically, this would allow us to avoid GAT-related bugs and restore the property.

I implemented this on [this branch](https://github.com/JoJoJet/bevy/commits/system-param-gats-trait), though I stopped after running into unexpected errors. This possibility is a work in progress.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:131]: This new definition is wildly easier to understand!

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:132]: I would like full doc comments on the methods and types here: now that the hierarchy is flatter this is actually possible to explain fairly directly.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:175]: This safety comment doesn't make sense to me. Only F is constrained to be read only here, and I don't see how this lines up with the safety requirements on SystemParam.

File: crates/bevy_render/src/extract_param.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/extract_param.rs:101]: The bounds were weakened here, but IMO correctly. ReadOnlySystemParam implies SystemParam.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (JoJoJet) [reply crates/bevy_ecs/src/system/system_param.rs:175]: Probably a rebase error, thanks for catching it.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_param.rs:175]: Ah, this was copied from above by mistake! It's correct there, but totally wrong here.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:129]: This could really use some more details.

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:128]: This doc link is broken (and wrong).

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:127]: This could use more pointers to where `system_meta` is stored (ideally a link, so we're notified if it gets moved).

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/system_param.rs:133]: As a reader, it's hard to follow why the Item associated type also needs to implement the SystemParam trait (and how that tower comes to an end).

File: crates/bevy_ecs/src/system/system_param.rs
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/system_param.rs:535]: Hunting down all mentions of states in comments will be a pain...

File: crates/bevy_ecs/src/system/system_param.rs
Comment (JoJoJet) [reply crates/bevy_ecs/src/system/system_param.rs:138]: Do you think it's clearer now?

**Comment by JoJoJet** [comment]
### 2. Lose the property

If we treat `::Item<'w, 's>` as an 'operation', then we want to make sure that performing the operation multiple times does the same thing as performing it once. Operations like this are known as 'idempotent'. Idempotency is a very nice property for any operation to have, but in this case it seems to have niche benefits. Throughout the entire engine and all of the tests and examples, the only code that relied on idempotency was this doctest:

```rust
pub fn pipe<AIn, Shared, BOut, A, AParam, AMarker, B, BParam, BMarker>(
    mut a: A,
    mut b: B,
) -> impl FnMut(In<AIn>, ParamSet<(SystemParamItem<AParam>, SystemParamItem<BParam>)>) -> BOut
where
    A: SystemParamFunction<AIn, Shared, AParam, AMarker>,
    B: SystemParamFunction<Shared, BOut, BParam, BMarker>,
    AParam: SystemParam,
    BParam: SystemParam,
```

Given how niche this is, I think that simply losing idempotency is a viable option.

**Comment by JoJoJet** [comment]
I was finally able to get the above example working. By simplifying the type of `ParamSet::Item<'w, 's>`, that example does not need `SystemParam::Item<'w, 's>` to be idempotent, and CI passes.

If we go in this direction, there is one open question: are we okay with losing the property `SystemParamItem<SystemParamItem<T>>` == `SystemParamItem<T>`? I am not aware of any use-cases for this property other than the aforementioned example, which itself is quite contrived (and no longer an issue).

File: crates/bevy_ecs/src/system/system_param.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_param.rs:137]: Oh that's so much clearer!

**Review by alice-i-cecile — COMMENTED** [review]
Thank you very much for adding full docs to the SystemParam trait! Those + simpler trait hierarchy makes this code way more approachable.

**Review by alice-i-cecile — COMMENTED** [review]
This has my approval, pending the restoring of the invariant mentioned in the first comment on this PR.

**Review by james7132 — APPROVED** [review]
Generally checks out to me, I really like how much this cleans up this area of bevy_ecs, though I am going to defer to @BoxyUwU for the lifetime shenanigans going on here.

File: crates/bevy_ecs/macros/src/lib.rs
Comment (JoJoJet) [reply crates/bevy_ecs/macros/src/lib.rs:433]: We no longer generate a struct for the state, we just use a tuple. I suppose we could just remove the anonymous scope entirely.

**Comment by JoJoJet** [comment]
Renamed `SystemParam::init` to `init_state`, to match the `WorldQuery` trait.

File: crates/bevy_ecs/macros/src/lib.rs
Comment (JoJoJet) [reply crates/bevy_ecs/macros/src/lib.rs:433]: I'm going to leave it as-is for now. Removing the anonymous scope would just add more churn, and its possible we might need to re-add it in the future.

**Comment by JoJoJet** [comment]
Fixed the merge conflicts, and did some minor syntax cleanups along the way.

**Comment by JoJoJet** [comment]
@alice-i-cecile can we consider this for the 0.10 milestone? If merged, I'd like it to land in the same release as #6865 so we don't break users multiple times.

**Review by alice-i-cecile — APPROVED** [review]
I'm happy with this: I've gone over it carefully and it appears to be correctly done. This is a substantial improvement to the clarity here. Fantastic migration guide in particular.

bors r+

**Comment by JonahPlusPlus** [comment]
Okay, saw this was merged. I'll start my system param PR now.

**Pull request closed without merging**
